apiserver/facades/agent/provisioner/provisioninginfo.go

```diff
@@ -51,22 +58,29 @@ func (api *ProvisionerAPI) ProvisioningInfo(args params.Entities) (params.Provis
 		}
 		machine, err := api.getMachine(canAccess, tag)
 		if err == nil {
-			result.Results[i].Result, err = api.getProvisioningInfo(machine, env)
+			result.Results[i].Result, err = api.getProvisioningInfo(machine, env, spacesIdsToProviderIds)
 		}
+
 		result.Results[i].Error = common.ServerError(err)
 	}
 	return result, nil
 }
 
-func (api *ProvisionerAPI) getProvisioningInfo(m *state.Machine, env environs.Environ) (*params.ProvisioningInfoV10, error) {
-	var err error
-	var result params.ProvisioningInfoV10
+func (api *ProvisionerAPI) getProvisioningInfo(m *state.Machine,
+	env environs.Environ,
+	spacesIdsToProviderIds map[string]string,
+) (*params.ProvisioningInfoV10, error) {
+	endpointBindings, err := api.machineEndpointBindings(m, spacesIdsToProviderIds)
```

Comment:
I'd rather not perpetuate the ambiguity of this down the call stack. When this is used in `machineSpaceTopology`, we won't know whether the bound spaces are names or provider IDs (we know they will be for MAAS, but I mean semantically expressed here for cogency).

I think we need to get the endpoint bindings (without conversion), and `AllSpaceInfos` up here, and pass both of those to `getProvisioningInfoBase` and `machineSpaceTopology`. In the former we can convert to provider ID/name; in the latter we will know that we have space IDs for the bindings.

Given that we already retrieved all the space info, we can use it in `subnetsAndZonesForSpace` instead of retrieving the subnet by ID and then its subnets in separate steps.